experimental/sql/tools/rel_opt.py

```diff
@@ -4,12 +4,33 @@
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 from xdsl.xdsl_opt_main import xDSLOptMain
+from io import IOBase
 
+from src.ibis_frontend import ibis_to_xdsl
 from dialects.ibis_dialect import Ibis
 
 
 class RelOptMain(xDSLOptMain):
 
+  def register_all_frontends(self):
+    super().register_all_frontends()
+
+    def parse_ibis(f: IOBase):
```

Comment:
This is still something I am not very proud of and it comes from the fact that this rel_opt tool wants to read one file and build an IR from that. I think it might make sense to extensively test using `pytest` further down the line (AFAIU, the same framework ibis uses for their test-suite).

Reply:
How about a while loop on `f.readline()` that filters out lines that `startswith(#)`? Similar to [this SO post](https://stackoverflow.com/a/6985913/651937) and potentially as a [generator](https://stackoverflow.com/a/519653/651937).

Reply:
Actually, if you prefix the `CHECK*` with ` #`, `eval`ing the whole file could be fine?

Reply:
Still, I don't like this approach a lot, since the query actually isn't generated form just the file, but from some local context within rel_opt as well...

Reply:
I am not sure I follow 100% but maybe you could add a fake test_harness op that would take l19-28 as a python string and interpret it on the fly.

Reply:
Maybe the context should be part of the test case? It could still be shortened slightly by moving some of it to some `test_utils` package or similar.

```python
import ibis
import pandas as pd

connection = ibis.pandas.connect(
  {"t": pd.DataFrame({"a": ["AS", "EU", "NA"]})})
table = connection.table('t')
```

But maybe that doesn't combine well with `eval`? I think that the root cause of this whole issue is the fact that Ibis isn't a language but an API...

Eventually, I think that we want to call `ibis_to_xdsl` from our own Ibis backend. But writing that could involve a serious amount of plumbing, so I wouldn't do it right now.

I am inclined to say to kick the can down the road: live it as is for now and worry about it later.

Reply:
Let me rephrase: Fundamentally, I feel that we are not really generating IR from a file like a *file*check test usually does, but rather from a file and a context. The exact schema of a table for example is part of that context. A linked part would be where to find column "a" in a tuple of table "t" (at position 0? or 1?) or where "t" actually is in memory. This is information that we will need at some point to actually be able to lower, but I am not sure whether this should be part of the IR of a query.

I would suggest leaving it as is and reopening this discussion once we start lowering to less abstract representations. Then we might be able to see more easily if this information should be part of the IR, or if this should be passed into the lowering pass as a context argument.

Reply:
OK, I think I understand. And I agree: let's leave it as is for now.

Reply:
sgtm!